File: lecture_1/problem_set/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int get_doubled_digits(long card);
int get_single_digits(long card);
int get_int_length(long num);
string get_issuer(long card);

int main(void)
{

    long card_num = get_long("Card number: ");

    string issuer = get_issuer(card_num);

    if ((get_doubled_digits(card_num) + get_single_digits(card_num)) % 10 != 0)
    {
        issuer = "INVALID";
    }

    printf("%s\n", issuer);
}

int get_doubled_digits(long card)
{
    int total = 0;
    while (card > 0)
    {
        long discard = card / 10;
        long double_digit = (discard % 10) * 2;
        if (double_digit >= 10)
        {
            total += (double_digit / 10) + (double_digit % 10);
        }
        else
        {
            total += double_digit;
        }
        card = card / 100;
    }
    return total;
}

int get_single_digits(long card)
{
    int total = 0;
    while (card > 0)
    {
        long single_digit = (card % 10);
        total += single_digit;
        card = card / 100;
    }
    return total;
}

int get_int_length(long num)
{
    int length = 1;
    while (num > 9)
    {
        length++;
        num = num / 10;
    }
    return length;
}

string get_issuer(long card)
{
    int card_length = get_int_length(card);
    if (card_length == 13 || card_length == 15 || card_length == 16)
    {
        int first_two = 0;
        if (card_length == 13)
        {
            first_two = card / 100000000000;
            if (first_two >= 40 && first_two <= 49)
            {
                return "VISA";
            }
            else
            {
                return "INVALID";
            }
        }
        else if (card_length == 15)
        {
            first_two = card / 10000000000000;
            if (first_two == 34 || first_two == 37)
            {
                return "AMEX";
            }
            else
            {
                return "INVALID";
            }
        }
        else
        {
            first_two = card / 100000000000000;
            if (first_two >= 40 && first_two <= 49)
            {
                return "VISA";
            }
            else if (first_two >= 51 && first_two <= 55)
            {
                return "MASTERCARD";
            }
            else
            {
                return "INVALID";
            }
        }
    }
    else
    {
        return "INVALID";
    }
}

File: week_6-python/pset/sentimental-credit/credit.py
import cs50
import re


def main():
    card = cs50.get_string("Card number: ")
    print(luhn_algo(card))


def luhn_algo(card_num):

    # By default, start with issuer being "INVALID"
    issuer = "INVALID"

    # Legitimate card numbers are only 16, 15, or 13 digits
    if re.fullmatch("\d{16}|\d{15}|\d{13}", card_num):
        # Convert to digits
        digits = [int(d) for d in card_num]

        # Iterate through every second digit in reverse and mutliply
        # Start at -2, stop at -1, step by -2
        for i in range(len(digits) - 2, -1, -2):
            digits[i] *= 2
            if digits[i] > 9:
                digits[i] -= 9

        # Calculate the sum
        total = sum(digits)

        # If total % 10 = 0, the number is valid. Get the issuer
        if total % 10 == 0:
            # Additional error checking included for number length, just to make sure
            # Mastercard is 16 digits and starts with 51, 52, 53, 54, or 55
            if re.fullmatch("\d{16}", card_num) and re.match(
                "51|52|53|54|55", card_num
            ):
                issuer = "MASTERCARD"
            # Visa is 13 or 16 digits and starts with 4
            elif re.fullmatch("\d{16}|\d{13}", card_num) and re.match("4", card_num):
                issuer = "VISA"
            # Amex is 15 digits and starts with 34 or 37
            elif re.fullmatch("\d{15}", card_num) and re.match("34|37", card_num):
                issuer = "AMEX"

    return issuer


main()
